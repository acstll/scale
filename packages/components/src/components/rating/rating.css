/**
* @prop --rating--container-display: Display of the rating container
* @prop --rating--star-emoji-display: Display of the rating star rating emoji
* @prop --rating--star-selected-color: Color of the rating star selected
* @prop --rating--star-selected-after-content: Content of the rating star selected after
* @prop --rating--emoji-selected-1-after-content: Content of the rating star selected 1 after
* @prop --rating--emoji-selected-2-after-content: Content of the rating star selected 2 after
* @prop --rating--emoji-selected-3-after-content: Content of the rating star selected 3 after
* @prop --rating--emoji-selected-4-after-content: Content of the rating star selected 4 after
* @prop --rating--emoji-selected-5-after-content: Content of the rating star selected 5 after

*/

.rating--container {
  display: var(--rating--container-display, block);
}

.rating--star,
.rating--emoji {
  display: var(--rating--star-emoji-display, inline);
  margin-right: var(--rating--star-emoji-margin-right, 6px);
  cursor: var(--rating--star-emoji-cursor, pointer);
  font-size: var(--rating--star-emoji-font-size, 24px);
  transition: var(--rating--star-emoji-transition, color 0.2s);
}

.rating--star:after {
  content: var(--rating--star-after-content, '\2606');
}
.rating--emoji:after {
  content: var(--rating--emoji-after-content, '\1F611');
}
.rating--star-selected-1,
.rating--star-selected-2,
.rating--star-selected-3,
.rating--star-selected-4,
.rating--star-selected-5 {
  color: var(--rating--star-selected-color, #f7ba2a);
}
.rating--star-selected-1:after,
.rating--star-selected-2:after,
.rating--star-selected-3:after,
.rating--star-selected-4:after,
.rating--star-selected-5:after {
  content: var(--rating--star-selected-after-content, '\2605');
}
.rating--emoji-selected-1:after {
  content: var(--rating--emoji-selected-1-after-content, '\1F61E');
}
.rating--emoji-selected-2:after {
  content: var(--rating--emoji-selected-2-after-content, '\1F615');
}
.rating--emoji-selected-3:after {
  content: var(--rating--emoji-selected-3-after-content, '\1F625');
}
.rating--emoji-selected-4:after {
  content: var(--rating--emoji-selected-4-after-content, '\1F60E');
}
.rating--emoji-selected-5:after {
  content: var(--rating--emoji-selected-5-after-content, '\1F60D');
}
