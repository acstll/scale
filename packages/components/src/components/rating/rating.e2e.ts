import { newE2EPage } from '@stencil/core/testing';

describe('t-rating', () => {
  it('renders', async () => {
    const page = await newE2EPage();
    await page.setContent('<t-rating></t-rating>');
    const element = await page.find('t-rating');
    expect(element).toHaveClass('hydrated');
  });
});
