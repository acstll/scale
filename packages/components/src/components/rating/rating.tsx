import { Component, h, Prop, State } from '@stencil/core';
import { CssClassMap } from '../../utils/utils';
import classNames from 'classnames';

@Component({
  tag: 't-rating',
  styleUrl: 'rating.css',
  shadow: true,
})
export class Rating {
  @Prop() public type?: string;
  @Prop() public range = [1, 2, 3, 4, 5];
  @Prop() public disabled?: boolean = false;

  @State() private selectedValue = null;
  @State() private preSelectValue = 0;
  @State() private selected = false;

  public handleMouseEnter = val => {
    this.preSelectValue = val;
  };

  public handleMouseLeave = val => {
    if (this.selectedValue) {
      this.preSelectValue = val;
    } else {
      this.preSelectValue = 0;
    }
  };

  public handleClick = val => {
    if (this.selectedValue !== val) {
      this.selectedValue = val;
      this.preSelectValue = val;
    } else {
      this.selectedValue = null;
      this.preSelectValue = 0;
    }
  };

  public render() {
    return (
      <div class="rating--container">
        {this.range.map(val => {
          this.selected = val <= this.preSelectValue;
          return (
            <div
              class={this.getCssClassMap()}
              onMouseEnter={() => this.handleMouseEnter(val)}
              onMouseLeave={() => this.handleMouseLeave(val)}
              onClick={() => this.handleClick(val)}
            />
          );
        })}
      </div>
    );
  }

  private getCssClassMap(): CssClassMap {
    return classNames(
      `rating--${this.type}`,
      this.selected && `rating--${this.type}-selected-${this.preSelectValue}`
    );
  }
}
