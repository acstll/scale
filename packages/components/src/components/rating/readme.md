# t-rating



<!-- Auto Generated Below -->


## Properties

| Property | Attribute | Description | Type     | Default     |
| -------- | --------- | ----------- | -------- | ----------- |
| `type`   | `type`    |             | `string` | `undefined` |


## CSS Custom Properties

| Name                                       | Description                                  |
| ------------------------------------------ | -------------------------------------------- |
| `--rating--container-display`              | Display of the rating container              |
| `--rating--emoji-after-content`            | Content of the rating emoji after            |
| `--rating--emoji-selected-1-after-content` | Content of the rating star selected 1 after  |
| `--rating--emoji-selected-2-after-content` | Content of the rating star selected 2 after  |
| `--rating--emoji-selected-3-after-content` | Content of the rating star selected 3 after  |
| `--rating--emoji-selected-4-after-content` | Content of the rating star selected 4 after  |
| `--rating--emoji-selected-5-after-content` | Content of the rating star selected 5 after  |
| `--rating--star-after-content`             | Content of the rating star after             |
| `--rating--star-emoji-cursor`              | Cursor of the rating star rating emoji       |
| `--rating--star-emoji-display`             | Display of the rating star rating emoji      |
| `--rating--star-emoji-font-size`           | Font size of the rating star rating emoji    |
| `--rating--star-emoji-margin-right`        | Margin right of the rating star rating emoji |
| `--rating--star-emoji-transition`          | Transition of the rating star staring emoji  |
| `--rating--star-selected-after-content`    | Content of the rating star selected after    |
| `--rating--star-selected-color`            | Color of the rating star selected            |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
