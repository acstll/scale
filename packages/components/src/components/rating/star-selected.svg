<svg width="24px" height="24px" viewBox="0 0 24 24" version="1.1">
    <g id="t-star" stroke="none" stroke-width="1" fill="none" fill-rule="evenodd">
        <polygon id="Star" stroke="#f7ba2a" fill="#f7ba2a" points="12 18.75 5.2404696 22.3036954 6.53142503 14.7768477 1.06285006 9.44630456 8.6202348 8.34815228 12 1.5 15.3797652 8.34815228 22.9371499 9.44630456 17.468575 14.7768477 18.7595304 22.3036954"></polygon>
    </g>
</svg>