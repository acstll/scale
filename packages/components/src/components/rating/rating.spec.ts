import { newSpecPage } from '@stencil/core/testing';
import { Rating } from './rating';

describe('Rating', () => {
  let element;
  beforeEach(async () => {
    element = new Rating();
  });

  it('should match snapshot', async () => {
    const page = await newSpecPage({
      components: [Rating],
      html: `<t-rating>/t-rating>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('should set preSelectValue when mouse entering', () => {
    element.preSelectValue = 0;
    element.handleMouseEnter(1);
    expect(element.preSelectValue).toBe(1);
  });

  it('should set new preSelectValue if there is a value set to selectedValue when mouse leaving', () => {
    element.selectedValue = 2;
    element.handleMouseLeave(5);
    expect(element.preSelectValue).toBe(2);
  });

  it('should set new preSelectValue to 0 if there is no value set to preSelectValue when mouse leaving', () => {
    element.selectedValue = 0;
    element.handleMouseLeave(5);
    expect(element.preSelectValue).toBe(0);
  });

  it('should set new selectedValue and preSelectValue if the selectedValue is not the same as param value when clicking', () => {
    element.selectedValue = 1;
    element.handleClick(5);
    expect(element.selectedValue).toBe(5);
    expect(element.preSelectValue).toBe(5);
  });

  it('should set new selectedValue to null and preSelectValue to 0 if the selectedValue is the same as param value when clicking', () => {
    element.selectedValue = 1;
    element.handleClick(1);
    expect(element.selectedValue).toBe(null);
    expect(element.preSelectValue).toBe(0);
  });

  it('should have class emoji when type value emoji is set', () => {
    element.type = 'emoji';
    expect(element.getCssClassMap()).toContain('rating__item--emoji');
  });
});
