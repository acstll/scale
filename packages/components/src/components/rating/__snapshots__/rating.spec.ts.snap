// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rating should match snapshot 1`] = `
<t-rating>
  <mock:shadow-root>
    <div class="rating__container">
      <div class="rating__item--undefined"></div>
      <div class="rating__item--undefined"></div>
      <div class="rating__item--undefined"></div>
      <div class="rating__item--undefined"></div>
      <div class="rating__item--undefined"></div>
    </div>
  </mock:shadow-root>
  /t-rating&gt;
</t-rating>
`;
